### PR TITLE
Rdm 4351 - Add new DynamicList basetype

### PR DESCRIPTION
**Pull request by mohammedanas**
https://tools.hmcts.net/jira/browse/RDM-4351

Currently we have the 'FixedList' base type which services use to configure a list of options that the user can pick from. This data is fixed as per the configuration.

With the introduction of 'DynamicList' base type services will be able to update the list of options in the list dynamically via the AboutToStart and/or MidEvent call backs

This story is to create the 'DynamicList' base type and add to the definition store

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ListFieldTypeParser.java [hunk 1]
```diff
@@ -81,6 +84,14 @@ public ListFieldTypeParser(ParseContext parseContext) {
         parseContext.addToAllTypes(fixedListType);
         result.addNew(fixedListType);
 
+        // Add as DynamicList with no list items as it will be dynamically added by the services
+        final FieldTypeEntity dynamicListType = new FieldTypeEntity();
+        dynamicListType.setBaseFieldType(fixedListBaseType);
```
Comment (mario-paniccia) [inline excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ListFieldTypeParser.java]: shouldn't this be:
`dynamicListType.setBaseFieldType(dynamicListBaseType);`
?

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ListFieldTypeParser.java [hunk 2]
```diff
@@ -81,6 +84,14 @@ public ListFieldTypeParser(ParseContext parseContext) {
         parseContext.addToAllTypes(fixedListType);
         result.addNew(fixedListType);
 
+        // Add as DynamicList with no list items as it will be dynamically added by the services
+        final FieldTypeEntity dynamicListType = new FieldTypeEntity();
```
Comment (mario-paniccia) [inline excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ListFieldTypeParser.java]: the ticket says:
```If a list has been been defined using 'FieldTypeParameter' when it is 'DynamicList' it configured list should be ignored
In this case do we really need this block of code here then?

File: application/src/test/java/uk/gov/hmcts/net/ccd/definition/store/excel/SpreadSheetImportTest.java [hunk 3]
```diff
@@ -282,11 +282,13 @@ private void assertFieldTypes() {
     }
 
     private void assertListFieldTypes(List<Map<String, Object>> fieldTypes) {
-
         assertThat(fieldTypes,
                    allOf(hasItem(allOf(hasColumn("jurisdiction_id", getIdForTestJurisdiction()),
                                        hasColumn("reference", "FixedList-marritalStatusEnum"),
                                        hasColumn("base_field_type_id", fieldTypesId.get("FixedList")))),
+                       hasItem(allOf(hasColumn("jurisdiction_id", getIdForTestJurisdiction()),
+                                       hasColumn("reference", "DynamicList-marritalStatusEnum"),
```
Comment (mario-paniccia) [inline application/src/test/java/uk/gov/hmcts/net/ccd/definition/store/excel/SpreadSheetImportTest.java]: dynamic list most of the times will be configured with empty FieldTypeParameter. I think we should have that kind of test here. 
Here we have tested FieldTypeParameter=marritalStatusEnum which is ok, we can leave it. But we must have also a test for FieldTypeParameter empty.

File: repository/src/main/resources/db/changelog/db.changelog-rdm-4351.xml [hunk 4]
```diff
@@ -0,0 +1,18 @@
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
```
Comment (mario-paniccia) [inline repository/src/main/resources/db/changelog/db.changelog-rdm-4351.xml]: we need to add this new field type to the ElasticSearch mappings too. Look at:
application.yml
After you do that, can you make sure you are able to import a spreadsheet with ElasticSearch turned on? If you try that at the moment, you will see
the import fails because on an ElasticSearch error. We need to fix that

File: application/src/test/java/uk/gov/hmcts/net/ccd/definition/store/excel/SpreadSheetImportTest.java [hunk 5]
```diff
@@ -282,11 +282,13 @@ private void assertFieldTypes() {
...
+                                       hasColumn("base_field_type_id", fieldTypesId.get("FixedList")))),
                          hasItem(allOf(hasColumn("jurisdiction_id", getIdForTestJurisdiction()),
                                        hasColumn("reference", "MultiSelectList-marritalStatusEnum"),
```
Comment (mario-paniccia) [inline application/src/test/java/uk/gov/hmcts/net/ccd/definition/store/excel/SpreadSheetImportTest.java]: furthermore, is it possible to add some more unit tests rather than relying mostly on this test here which is an end to end test

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ListFieldTypeParser.java (hunk 1 above)
Comment (mohammedanas) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ListFieldTypeParser.java]: correct thanks for spotting it

File: repository/src/main/resources/db/changelog/db.changelog-rdm-4351.xml (hunk 4 above)
Comment (mohammedanas) [reply repository/src/main/resources/db/changelog/db.changelog-rdm-4351.xml]: Added them

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ListFieldTypeParser.java (hunk 2 above)
Comment (mohammedanas) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ListFieldTypeParser.java]: Removed check for fieldtypeparameter changes for DynamicList

File: application/src/test/java/uk/gov/hmcts/net/ccd/definition/store/excel/SpreadSheetImportTest.java (hunk 3 above)
Comment (mohammedanas) [reply application/src/test/java/uk/gov/hmcts/net/ccd/definition/store/excel/SpreadSheetImportTest.java]: Yes tested with both cases and added to IT excel sheet

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ListFieldTypeParser.java [hunk 6]
```diff
@@ -28,18 +29,20 @@
...
     private final FieldTypeEntity multiListBaseType;
+    private final FieldTypeEntity dynamicListBaseType;
 
     public ListFieldTypeParser(ParseContext parseContext) {
         this.parseContext = parseContext;
         fixedListBaseType = parseContext.getBaseType(FIXED_LIST_TYPE).orElseThrow(() -> new InvalidImportException(NO_BASE_TYPE_FOUND + FIXED_LIST_TYPE));
         fixedRadioListBaseType = parseContext.getBaseType(FIXED_RADIO_LIST_TYPE).orElseThrow(() -> new InvalidImportException(NO_BASE_TYPE_FOUND + FIXED_RADIO_LIST_TYPE));
         multiListBaseType = parseContext.getBaseType(MULTI_LIST_TYPE).orElseThrow(() -> new InvalidImportException(NO_BASE_TYPE_FOUND + MULTI_LIST_TYPE));
-
+        dynamicListBaseType = parseContext.getBaseType(DYNAMIC_LIST_TYPE).orElseThrow(() -> new InvalidImportException(NO_BASE_TYPE_FOUND + DYNAMIC_LIST_TYPE));
```
Comment (mario-paniccia) [inline excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ListFieldTypeParser.java]: the ticket says: 
```If a list has been been defined using 'FieldTypeParameter' when it is 'DynamicList' it configured list should be ignored```
given this requirement, can we just remove this line here? Because we have to ignore this info anyway

Comment (mario-paniccia) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ListFieldTypeParser.java]: perhaps removing that line will fix RDM-4634?
